File: sudoq-app/sudoqmodel/src/main/java/de/sudoq/model/solvingAssistant/SolveActionLastDigit.java
package de.sudoq.model.solvingAssistant;

import java.util.List;

import de.sudoq.model.sudoku.Position;
import de.sudoq.model.sudoku.Utils;

/**
 * Created by timo on 28.04.16.
 */
public class SolveActionLastDigit extends SolveAction {
    public SolveActionLastDigit(Position target, int solution, List<Position> list){
        Position rwPosition = Utils.positionToRealWorld(target);
        int      rwSolution = Utils.symbolToRealWorld(solution);
        /* are all x values the same? */
        int xPo = list.get(0).getX();
        boolean xSame = true;
        for (Position p: list)
            if (p.getX() != xPo) {
                xSame = false;
                break;
            }
        /* are all Y values the same? */
        int yPo = list.get(0).getY();
        boolean ySame = true;
        for (Position p: list)
            if (p.getY() != yPo) {
                ySame = false;
                break;
            }

        String structure = xSame ? "column":
                           ySame ? "row":
                                   "block";

        message = "Field at row "+ rwPosition.getX() +", column "+ rwPosition.getY() + " must be " + rwSolution + ", because all other fields in it's "+structure+" are occupied";
    }
}

File: sudoq-app/sudoqmodel/src/main/java/de/sudoq/model/solvingAssistant/SolveActionLastDigit.kt
package de.sudoq.model.solvingAssistant

import de.sudoq.model.sudoku.Position
import de.sudoq.model.sudoku.Utils.positionToRealWorld
import de.sudoq.model.sudoku.Utils.symbolToRealWorld

class SolveActionLastDigit(target: Position, solution: Int, list: List<Position>) : SolveAction() {
    init {
        val rwPosition = positionToRealWorld(target)
        val rwSolution = symbolToRealWorld(solution)
        /* are all x values the same? */
        val xPo = list[0].x
        var xSame = true
        for (p in list) if (p.x != xPo) {
            xSame = false
            break
        }
        /* are all Y values the same? */
        val yPo = list[0].y
        var ySame = true
        for (p in list) if (p.y != yPo) {
            ySame = false
            break
        }
        val structure = if (xSame) "column" else if (ySame) "row" else "block"
        message = "Field at row " + rwPosition.x + ", column " + rwPosition.y + " must be " + rwSolution + ", because all other fields in it's " + structure + " are occupied"
    }
}